ui/mirage/factories/agent.js

```diff
@@ -8,20 +8,51 @@ const AGENT_STATUSES = ['alive', 'leaving', 'left', 'failed'];
 
 export default Factory.extend({
   id: i => (i / 100 >= 1 ? `${UUIDS[i]}-${i}` : UUIDS[i]),
-  name: () => `nomad@${faker.random.boolean() ? faker.internet.ip() : faker.internet.ipv6()}`,
+  config: makeConfig,
 
-  status: () => faker.helpers.randomize(AGENT_STATUSES),
-  serfPort: () => faker.random.number({ min: 4000, max: 4999 }),
-
-  address() {
-    return this.name.split('@')[1];
-  },
-
-  tags() {
-    const rpcPortCandidate = faker.random.number({ min: 4000, max: 4999 });
+  member() {
     return {
-      port: rpcPortCandidate === this.serfPort ? rpcPortCandidate + 1 : rpcPortCandidate,
-      dc: faker.helpers.randomize(DATACENTERS),
+      name: generateName(),
+      status: faker.helpers.randomize(AGENT_STATUSES),
+      serfPort: faker.random.number({ min: 4000, max: 4999 }),
     };
   },
+
+  afterCreate(agent) {
+    const address = generateAddress.bind(agent);
+    const tags = generateTags.bind(agent);
```

Comment:
These are computed properties based on other properties of the member object. Unsure if `afterCreate` was the best option. My next best idea was adding some init function to the member property and then doing something like this. I saw it on StackOverflow but then that pattern wasn't used in the codebase and afterCreate seemed like better documented pattern in Ember Mirage.

```
{
  x: string,
  y: boolean,
  init: () => this.y ? this.x : 'haha yeah'
}.init();
```

Reply:
Why did these have to be moved to an `afterCreate`?  According to the [docs](https://www.ember-cli-mirage.com/versions/v0.4.x/factories/) you can reference other attributes dynamically, and that's how it was implemented before.

Reply:
I still don't know the answer about why. I believe its because the reference to `this`. Let me get you a better answer later.

Reply:
I need to have a better look at this, but right now I'm not sure what the reason is for binding. You could just pass the `agent` data in as an argument instead of binding and using `this`. Someone might look at those functions and think, "yeah ok, but what is `this` here", using `generateName(agent)` comments the code as you go so I know that the data being read is something to do with `agent`, `this` doesn't give me that context.

Also, at first glance, you only read one piece of data of the `agent` data in each function, you could just pass the single piece of data in, or even maybe (again I'd have to look closer) just do it inline.

All very stylistic this and not super important I know, but I think `bind` usage is uncommon overall in project, and trying to channel other engs on the project I would guess they suggest not using it here - but I could be wrong.

I'll look a little deeper in a sec, but hope that makes sense.

Reply:
This situation is a little different since, we're not working with attributes of the class. In order to access `name` we'd have to access `this.member.name` which is a circular reference.

I think we'd normally have `Member` be its own entity in a database that's related to `Agent` and we wouldn't be writing it as an object here. And the docs do mention that when you're working with a relationship that you should use `trait` or `afterCreate`.

I'm not sure how else we can get around the `this` reference issue.